util/EIP-712/src/encode.rs

```diff
@@ -0,0 +1,356 @@
+					continue;
+				}
+				types.insert(&*field.type_);
+			}
+		}
+	}
+}
+
+fn encode_type(message_type: &str, message_types: &MessageTypes) -> Result<String> {
+	let deps = {
+		let mut temp = build_dependencies(message_type, message_types).ok_or_else(|| ErrorKind::NonExistentType)?;
+		temp.remove(message_type);
+		let mut temp = temp.into_iter().collect::<Vec<_>>();
+		(&mut temp[..]).sort_unstable();
+		temp.insert(0, message_type);
+		temp
+	};
+
+	let encoded = deps
+		.into_iter()
+		.filter_map(|dep| {
+			message_types.get(dep).map(|field_types| {
+				let types = field_types
+					.iter()
+					.map(|value| format!("{} {}", value.type_, value.name))
+					.join(",");
+				return format!("{}({})", dep, types);
+			})
+		})
+		.collect::<Vec<_>>()
+		.concat();
+	Ok(encoded)
+}
+
+fn type_hash(message_type: &str, typed_data: &MessageTypes) -> Result<H256> {
+	Ok(keccak(encode_type(message_type, typed_data)?))
+}
+
+fn encode_data(
+	parser: &Parser,
+	message_type: &Type,
+	message_types: &MessageTypes,
+	value: &Value,
+	field_name: Option<&str>
+) -> Result<Vec<u8>>
+{
+	let encoded = match message_type {
+		Type::Array(array_type) => {
+			let mut items = vec![];
+
+			for item in value.as_array().ok_or_else(|| serde_error("array", field_name))? {
+				let mut encoded = encode_data(parser, &*array_type, &message_types, item, field_name)?;
+				items.append(&mut encoded);
+			}
+
+			keccak(items).0.to_vec()
+		}
+
+		Type::Custom(ref ident) if message_types.get(&*ident).is_some() => {
+			let type_hash = (&type_hash(ident, &message_types)?).to_vec();
+			let mut tokens = encode(&[EthAbiToken::FixedBytes(type_hash)]);
+
+			for field in message_types.get(ident).expect("Already checked in match guard; qed") {
+				let value = &value[&field.name];
+				let type_ = parser.parse_type(&*field.type_)?;
+				let mut encoded = encode_data(parser, &type_, &message_types, &value, Some(&*field.name))?;
+				tokens.append(&mut encoded);
+			}
+
+			keccak(tokens).0.to_vec()
+		}
+
+		Type::Bytes(size) if *size == 32 => {
+			let string = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if string.len() < 2 {
+				return Err(ErrorKind::HexParseError(
+					format!("Expected a 0x-prefixed string of even length, found {} length string", string.len()))
+				)?
+			}
+			let string = string.get(2..).expect("line 188 checks for length; qed");
+			let bytes = H256::from_str(string).map_err(|err| ErrorKind::HexParseError(format!("{}", err)))?;
+			let hash = (&keccak(&bytes)).to_vec();
+			encode(&[EthAbiToken::FixedBytes(hash)])
+		}
+
+		Type::Bytes(size) if *size < 32 => {
+			let string = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if string.len() < 2 {
+				return Err(ErrorKind::HexParseError(
+					format!("Expected a 0x-prefixed string of even length, found {} length string", string.len()))
+				)?
+			}
+			let string = string.get(2..).expect("line 200 checks for length; qed");
```

Comment:
* Why not just `[2..]` (and in other places)?
* Can we reuse code with case `if *size == 32`?